source/lac/sparse_direct.cc

```diff
@@ -1004,49 +1003,54 @@ SparseDirectMUMPS::initialize(const Matrix &matrix)
 {
   // Initialize MUMPS instance:
   initialize_matrix(matrix);
+
   // Start factorization
   id.job = 4;
   dmumps_c(&id);
 }
 
 void
-SparseDirectMUMPS::solve(Vector<double> &vector)
+SparseDirectMUMPS::vmult(Vector<double> &dst, const Vector<double> &src) const
 {
-  // TODO: this could be implemented similar to SparseDirectUMFPACK where
-  //  the given vector will be used as the RHS. Sadly, there is no easy
-  //  way to do this without breaking the interface.
-
-  // Check that the solver has been initialized by the routine above:
-  Assert(initialize_called == true, ExcNotInitialized());
-
   // and that the matrix has at least one nonzero element:
   Assert(nz != 0, ExcNotInitialized());
 
+  Assert(n == dst.size(), ExcMessage("Destination vector has the wrong size."));
+  Assert(n == src.size(), ExcMessage("Source vector has the wrong size."));
+
+  // Hand over right-hand side
+  copy_rhs_to_mumps(src);
+
   // Start solver
-  id.job = 6; // 6 = analysis, factorization, and solve
+  id.job = 3;
   dmumps_c(&id);
-  copy_solution(vector);
+  copy_solution(dst);
 }
 
 void
-SparseDirectMUMPS::vmult(Vector<double> &dst, const Vector<double> &src)
+SparseDirectMUMPS::Tvmult(Vector<double> &dst, const Vector<double> &src) const
 {
-  // Check that the solver has been initialized by the routine above:
-  Assert(initialize_called == true, ExcNotInitialized());
-
-  // and that the matrix has at least one nonzero element:
+  // The matrix has at least one nonzero element:
   Assert(nz != 0, ExcNotInitialized());
 
   Assert(n == dst.size(), ExcMessage("Destination vector has the wrong size."));
   Assert(n == src.size(), ExcMessage("Source vector has the wrong size."));
 
+  id.icntl[8] = 2; // transpose
   // Hand over right-hand side
   copy_rhs_to_mumps(src);
 
   // Start solver
   id.job = 3;
   dmumps_c(&id);
```

Comment:
I continue to be tickled by the fact that the MUMPS folks chose a system by which their public interface consists of *exactly one* function. I also question their software design skills.